api/envoy/config/filter/network/find_and_replace/v2alpha1/find_and_replace.proto

```diff
@@ -0,0 +1,39 @@
+syntax = "proto3";
+
+package envoy.config.filter.network.find_and_replace.v2alpha1;
+option go_package = "v2alpha1";
+
+import "validate/validate.proto";
+import "gogoproto/gogo.proto";
+
+// [#protodoc-title: Find and Replace]
+// Find and Replace Network Filter :ref:`configuration overview
+// <config_network_filters_find_and_replace>`.
+
+// Find and Replace network filter config.
+//
+message FindAndReplace {
+  // Optional string to find in data from the downstream side (read filter).
+  // Finding the string is currently only supported from the very beginning of the data on a given
+  // connection, meaning the string has to match the beginning of the data. If empty (or not
+  // specified), then :ref:`input_rewrite_to
+  // <envoy_api_field_config.filter.network.find_and_replace.v2alpha1.FindAndReplace.input_rewrite_to>`
+  // effectively becomes an insertion instead of a replacement string. Defaults to empty.
+  string input_rewrite_from = 1;
```

Comment:
String matcher here? Same on the match below?

Also, to better future proof this, can we think about logically breaking this up into multiple different messages? For example:
1) Input matcher
2) Input translation
3) Output matcher
4) Output translation

I think we can then have common matching and translation constructs for both input and output, possible with some well placed oneofs to make it clear that we can add additional matching options in the future?

Reply:
I think I need some more guidance here. The issue I'm facing is that I don't know how to reasonably use the StringMatcher in conjunction with `Buffer::Instance::search()`. I guess I could utilize `Buffer::Instance::toString()` and call `Matchers::StringMatcher::match()` with that, but how would that efficiently work for any matching mode besides `prefix`?

For the `exact` match: To truly do an exact match I'd have to wait for all incoming data ever., which defeats the purpose here.

The `suffix` and `regex` matches have a similar issue: How do I know when to stop buffering and execute the comparison, since I don't know the length of the search string?

`prefix` does make sense, since I know the string length (which is the current functionality of this PR).

For example, let's assume incoming data from downstream is split across 2 packets. The first packet contains `abc` and the second contains `def`.
- `prefix` StringMatcher using `abcd` would work after receiving the second packet (likely the second `onData()` invocation).
- `exact` StringMatcher using `abcd` would never match, since both `onData()` calls never just contain `abcd`, and even if they did, then it would still be wrong, since it is an incoming data stream and I don't know how many more `onData()` invocations will happen.
- `suffix` and `regex` StringMatcher using `abcd` have nearly the same issue as `exact`: I don't know when the incoming data stream ends. (And currently it seems as if we're using `regex_match`, which works on the whole string, so it is facing the same issues as `exact`)

For any mode besides `prefix` we would have to know how much incoming data to wait for.

Does that make any sense or am I completely off here?